### PR TITLE
feat(dataset): trajectory-to-env pipeline for fine-tuning dataset construction

### LINKED ISSUE #3718: feat: trajectory-to-env pipeline for fine-tuning dataset construction from gptme sessions
## Summary

gptme sessions produce thousands of agent trajectories per day. A recent paper — [Terminal-Universe (arXiv:2609.04148)](https://arxiv.org/abs/2609.04148) — demonstrates that agent trajectories can be reconstructed into executable training environments at scale, yielding substantial benchmark gains (+11-14pt on Qwen3.5-27B fine-tuning from 37k environments).

I prototyped this pipeline against Bob's gptme corpus and found **58% of sessions are convertible** into training environments. At ~110k sessions/year, that's ~21k environments/year from the existing corpus alone.

## The core insight: commit-message session-ID attribution

gptme sessions naturally embed session IDs in commit messages, e.g.:

```
chore(journal): session c876 — refactor done
fix(tool): handle edge case (a1b2)
```

This pattern means we can recover the exact git state before/after each session **without replaying the trajectory**. Given a session ID, we can:
1. Find all commits where the message contains the session ID (e.g. `(c876)`)
2. The commit just before the first session commit is the `entry_commit` (workspace state when the session started)
3. The session commits are the `solution_commits`

This is a much cheaper recovery path than Terminal-Universe's general file-op replay — we get it for free from gptme's natural commit hygiene.

## What a training environment looks like

Each environment from the pipeline:
```json
{
  "session_id": "c876",
  "entry_commit": "abc123",
  "solution_commits": ["def456", "ghi789"],
  "files_changed": ["scripts/foo.py", "tests/test_foo.py"],
  "task_description": "CASCADE: fix tool edge case",
  "cascade_intent": "Fix edge case in bash tool output handling",
  "category": "code",
  "outcome": "productive",
  "tool_call_counts": {"Bash": 12, "Read": 8, "Edit": 4},
  "model": "claude-sonnet-4-6",
  "duration_seconds": 847.0
}
```

Given this, a fine-tuning pipeline can:
1. `git checkout entry_commit` — restore workspace to session start state
2. Inject `task_description` as the user prompt
3. Replay or generate the solution using `solution_commits` as ground truth
4. Apply Terminal-Universe breadth synthesis (generate 3-5 new tasks from the same workspace state) to multiply training signal

## Prototype

I have a working prototype at [`scripts/research/trajectory_to_env.py`](https://github.com/ErikBjare/bob/blob/master/scripts/research/trajectory_to_env.py) (private repo) that:
- Scans Claude Code session records (`~/.claude/projects/`)
- Finds associated git commits via session ID pattern matching
- Outputs `TaskEnvironment` JSONL records suitable for fine-tuning pipelines
- Reports corpus statistics (58% yield on last 500 sessions)

## Proposed integration with gptme

There are a few natural places this could live:

1. **`gptme/eval/`** — as a dataset builder that exports gptme-trajectory-derived environments in the same format as existing evals
2. **`gptme/benchmark/`** — as a self-improvement benchmark: does fine-tuning on gptme-derived environments improve gptme task performance?
3. **`gptme/data/`** — a new subpackage for dataset construction utilities, separate from eval

The terminal-universe paper uses environments for fine-tuning Qwen, but the gptme-specific value proposition is different: **using gptme's own session trajectories to train better gptme agents**, closing the meta-factory loop (see [GOALS.md: Self-Improvement](https://github.com/ErikBjare/bob/blob/master/GOALS.md)).

## Questions for maintainers

1. Is there interest in upstreaming this into gptme proper? The commit-message attribution trick is gptme-specific but the `TaskEnvironment` schema could be a shared interface.
2. Would the preferred shape be a CLI (`gptme dataset build-from-trajectories`) or a library function (`gptme.data.trajectory_to_env`)?
3. Are there privacy considerations for open-sourcing trajectory-derived environments? (The current corpus is private; an open-source version would need filtering.)

## References

- Terminal-Universe paper: https://arxiv.org/abs/2609.04148
- Research note (internal): `knowledge/research/2026-09-04-swe-gate-and-terminal-universe-agent-eval.md`
- Prototype (private): `scripts/research/trajectory_to_env.py`
- Related: Idea #5449 in the strategic backlog

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Implements the trajectory-to-env pipeline proposed in #3718.

Adds a `gptme.dataset` subpackage that converts gptme session trajectories
into `TaskEnvironment` JSONL records suitable for fine-tuning pipelines.

### Core insight

gptme sessions embed session IDs in commit messages:
```
chore(journal): session c876 — refactor done
fix(tool): handle edge case (a1b2)
```

This lets us recover the git state before/after each session **without replaying the trajectory** — a cheap alternative to the general file-op replay approach in the Terminal-Universe paper (arXiv:2609.04148).

### New modules

| File | Purpose |
|------|---------|
| `gptme/dataset/__init__.py` | Public API: `TaskEnvironment`, `extract_environments`, `scan_sessions` |
| `gptme/dataset/trajectory_to_env.py` | Core pipeline: session scanning, commit attribution, JSONL export |
| `gptme/cli/cmd_dataset.py` | CLI: `gptme-util dataset stats/export` (lazy-loaded via util dispatch) |
| `tests/test_dataset.py` | 26 offline unit + integration tests (no network, no real gptme logs) |

### Usage

```bash
# Corpus statistics
gptme-util dataset stats --repo /path/to/repo --limit 500

# Export TaskEnvironment JSONL
gptme-util dataset export --repo /path/to/repo -n 200 -o envs.jsonl

# Library usage
from gptme.dataset import extract_environments
envs = list(extract_environments(repo_path=Path('/path/to/repo'), limit=500))
```

### TaskEnvironment schema

```json
{
  "session_id": "c876",
  "entry_commit": "abc123",
  "solution_commits": ["def456", "ghi789"],
  "files_changed": ["scripts/foo.py", "tests/test_foo.py"],
  "task_description": "CASCADE: fix tool edge case",
  "category": "code",
  "outcome": "productive",
  "tool_call_counts": {"shell": 12, "python": 3},
  "model": "claude-sonnet-4-6",
  "duration_seconds": 847.0
}
```

Closes #3718

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
## Review guide (auto-posted)

Structured warm-up for reviewers — **what changed**, **what to run**, **where to look**.

### Key files
- `gptme/dataset/trajectory_to_env.py` (+457/-0, Δ457)
- `tests/test_dataset.py` (+396/-0, Δ396)
- `gptme/cli/cmd_dataset.py` (+182/-0, Δ182)
- `gptme/dataset/__init__.py` (+13/-0, Δ13)
- `gptme/cli/util.py` (+1/-0, Δ1)
- `pyproject.toml` (+1/-0, Δ1)

### Suggested verification
- `uv run pytest tests/test_dataset.py -q`
- `make typecheck`

### Known risks / watch points
- No automatic high-risk tags; use file list + diff for judgment.

### Suggested review focus
- Confirm behavior matches the PR description acceptance criteria.
- Skim the largest diffs first (listed above).
- If CI is green, spot-check the highest-risk paths called out here.

_Generated by `scripts/github/pr-warmup-review-guide.py` for gptme/gptme#3719._

**Comment by TimeToBuildBob** [comment]
## 🤖 AI code review

Adds a new `gptme/cli/cmd_dataset.py` exposing `gptme-util dataset stats` and `export` subcommands that delegate to the `gptme.dataset.trajectory_to_env` helpers and echo statistics or JSONL output; the PR also adds offline tests covering environment extraction, JSONL round-trip, session scanning filters, and CLI error behavior.

### Confidence Score: 4/5 — P2 findings only

<details><summary><b>1 finding</b> · ⚠️ **1** P2</summary>

**⚠️ P2 medium** — `gptme/cli/cmd_dataset.py:79`

In `stats_cmd`, the `except ValueError` branch (line 78-79) echoes the error but neither returns, re-raises, nor sets `result`, so control flows past the try/except. `result` is only bound inside the `try`, so on any scan failure (e.g. `--repo` not a git repository, which `corpus_stats` raises as ValueError per `test_cli_stats_no_repo`) the code after the handler runs with `result` unbound and raises `UnboundLocalError: local variable 'result' referenced before assignment`, surfacing a Python traceback instead of the intended clean error. If the tail never touches `result`, the command instead exits 0, reporting success for a failed scan. Either way the CLI does not yield the clean non-zero error path the stats command intends; `test_cli_stats_no_repo` passes only because the substring `Error` appears inside `UnboundLocalError` in the traceback.

```suggestion
click.echo(f"Error: {exc}", err=True)
        sys.exit(1)
```

<sub>**How this was verified:** The visible diff ends at this handler; it does not rebind `result` or terminate, while `test_cli_stats_no_repo` requires a non-zero exit, which can only come from later code raising on the unbound `result`.</sub>

</details>

<details><summary><b>1 existing-module overlap (informational, not scored)</b></summary>

_This is a cheap file-index scan, not a merge blocker. A human decides whether to redirect, extend the existing module, or intentionally parallel-implement._

**⚠️ P2 medium · ℹ️ non-blocking** — `gptme/dataset/trajectory_to_env.py:1`

This appears to overlap with `gptme/gptme-contrib:packages/gptme-rag` which already provides RAG: document indexing, semantic/lexical search, persistence. Matched signals: jsonl, knowledge, search. Informational only — not a blocker. A human decides whether to redirect, extend the existing module, or intentionally parallel-implement.

<sub>**How this was verified:** cheap file-index scan (L0.5 existing-module overlap)</sub>

</details>

<details><summary><b>Files changed (2)</b> — the diff as I read it</summary>

- `gptme/cli/cmd_dataset.py` — New click group `dataset` with `stats` and `export` subcommands that call `corpus_stats` and the `trajectory_to_env` helpers, printing stats or JSONL to stdout/stderr.
- `tests/test_dataset.py` — Adds offline tests for environment extraction, commit attribution, JSONL round-trip, session scanning filters, and CLI stats/export error paths.

<sub>Also changed, not described above: `gptme/cli/util.py`, `gptme/dataset/__init__.py`, `gptme/dataset/trajectory_to_env.py`, `pyproject.toml`</sub>

</details>

<details><summary>Previous review passes</summary>

| commit | score | findings | engine | when |
| --- | --- | --- | --- | --- |
| `e287838687b3` | 4/5 | 1 | llm | 2026-09-05 08:26 UTC |

</details>

<sub>Reviewed `5fc470dd2397` · `openrouter/deepseek/deepseek-v4-flash-0731` · `llm` engine · 841s · [about this reviewer](https://github.com/gptme/gptme-contrib/blob/master/docs/ai-reviewer.md)</sub>

<details><summary><sub>Maintainer commands</sub></summary>

<sub>`@TimeToBuildBob review` (own line) — fresh review · `@TimeToBuildBob fix` — a worker acts on the findings. Once per comment; 👀 = received.</sub>

</details>

File: gptme/cli/cmd_dataset.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+    if as_json:
+        click.echo(json.dumps(result, indent=2))
+        return
+
+    click.echo(f"Sessions scanned:   {result['scanned']}")
+    click.echo(f"Convertible:        {result['convertible']}")
+    click.echo(f"Yield:              {result['yield_pct']}%")
+    if result["category_counts"]:
+        click.echo("Categories:")
```
Comment (TimeToBuildBob) [inline gptme/cli/cmd_dataset.py:88]: ⚠️ **P2** — **P2 — `export_cmd` doesn't catch errors raised at iteration: traceback + partial output** · `gptme/cli/cmd_dataset.py` (the `export_cmd` try/except around `extract_environments`)

`extract_environments(...)` is a generator, so the `except ValueError` that wraps the call guards nothing: the body (including the “not a git repo” check) only runs at iteration inside `for env in env_iter:`. A `ValueError` raised there (e.g. `dataset export --output dataset.jsonl --repo not-a-repo`) escapes as an unhandled traceback, and any already-written output file is left as a corrupt/partial JSONL that gets presented as a completed export.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Addressed all Greptile P1/P2 findings in 5fc470dd2:

**P1 — Normal sessions skipped**: Removed hex-suffix requirement. `_attribution_ids()` now extracts the short hex only as a fallback — full conv IDs like `2026-01-01-fix-utils` work via `Git-Session-Id` trailer matching.

**P1 — Unrelated commits match**: Switched `git log` from `--all` to `HEAD`-only. Loose `--grep` replaced by structured `%H%x1f%B%x1e` format with exact `_SESSION_MARKER_TEMPLATE` + `Git-Session-Id` trailer verification.

**P1 — Merge changes omitted**: Replaced `git diff-tree --no-commit-id -r` with `git diff --name-only sha^ sha`, which picks up merge changes by comparing against the first parent.

**P1 — Validation escapes error handler**: Restructured `export_cmd` to wrap the generator iteration loop inside the `try/except`. Added atomic temp-file write (renamed on success, cleaned up on failure) to prevent partial output files.

**P1 — Repository subdirectories rejected**: Replaced `.git` existence check with `_resolve_repo()` using `git rev-parse --show-toplevel`, so running from any subdirectory works.

**P2 — Nonpositive min_commits crashes export**: Added `click.IntRange(min=1)` on the CLI option and a `ValueError` guard in `extract_environments()`.

36 tests pass, mypy clean.

File: gptme/cli/cmd_dataset.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    from ..dataset.trajectory_to_env import corpus_stats
+
+    try:
+        result = corpus_stats(
+            repo_path=repo_path.resolve(),
+            logs_dir=logs_dir,
+            limit=limit,
+        )
+    except ValueError as exc:
+        click.echo(f"Error: {exc}", err=True)
```
Comment (TimeToBuildBob) [inline gptme/cli/cmd_dataset.py:79]: ⚠️ **P2** — In `stats_cmd`, the `except ValueError` branch (line 78-79) echoes the error but neither returns, re-raises, nor sets `result`, so control flows past the try/except. `result` is only bound inside the `try`, so on any scan failure (e.g. `--repo` not a git repository, which `corpus_stats` raises as ValueError per `test_cli_stats_no_repo`) the code after the handler runs with `result` unbound and raises `UnboundLocalError: local variable 'result' referenced before assignment`, surfacing a Python traceback instead of the intended clean error. If the tail never touches `result`, the command instead exits 0, reporting success for a failed scan. Either way the CLI does not yield the clean non-zero error path the stats command intends; `test_cli_stats_no_repo` passes only because the substring `Error` appears inside `UnboundLocalError` in the traceback.

Comment (TimeToBuildBob) [reply gptme/cli/cmd_dataset.py:79]: Fixed in `5fc470dd2397` (fix(dataset): address all Greptile P1/P2 findings).

The `except ValueError` branch now calls `sys.exit(1)` immediately after the error echo, so control never falls through to the unbound `result` reference:

```python
except ValueError as exc:
    click.echo(f"Error: {exc}", err=True)
    sys.exit(1)
```

`test_cli_stats_no_repo` still passes and exercises this path directly.
